File: ts/src/scenarios/negotiation-spec.ts
import { z } from "zod";
import { SimulationActionSpecSchema } from "./simulation-spec.js";

export const HiddenPreferencesSchema = z.object({
  priorities: z.record(z.number()),
  reservationValue: z.number(),
  aspirationValue: z.number(),
  batnaDescription: z.string().min(1),
});

export const NegotiationSpecSchema = z.object({
  description: z.string().min(1),
  environmentDescription: z.string().min(1),
  initialStateDescription: z.string().min(1),
  hiddenPreferences: HiddenPreferencesSchema,
  maxRounds: z.number().int().min(2).max(10),
  successCriteria: z.array(z.string().min(1)).min(1),
  failureModes: z.array(z.string().min(1)).default([]),
  actions: z.array(SimulationActionSpecSchema).min(2),
  maxSteps: z.number().int().nonnegative().default(0),
});

export type HiddenPreferences = z.infer<typeof HiddenPreferencesSchema>;
export type NegotiationSpec = z.infer<typeof NegotiationSpecSchema>;

export function parseRawNegotiationSpec(data: Record<string, unknown>): NegotiationSpec {
  const parsed = NegotiationSpecSchema.parse({
    description: data.description,
    environmentDescription: data.environment_description,
    initialStateDescription: data.initial_state_description,
    hiddenPreferences: (() => {
      const raw = data.hidden_preferences as Record<string, unknown>;
      return {
        priorities: raw.priorities,
        reservationValue: raw.reservation_value,
        aspirationValue: raw.aspiration_value,
        batnaDescription: raw.batna_description,
      };
    })(),
    maxRounds: data.max_rounds,
    successCriteria: data.success_criteria,
    failureModes: data.failure_modes ?? [],
    actions: data.actions,
    maxSteps: data.max_steps ?? 0,
  });
  return {
    ...parsed,
    maxSteps: parsed.maxSteps > 0 ? parsed.maxSteps : Math.max(parsed.maxRounds * 2, 4),
  };
}

File: autocontext/src/autocontext/scenarios/custom/negotiation_spec.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Any

from autocontext.scenarios.custom.simulation_spec import SimulationActionSpecModel


@dataclass(slots=True)
class NegotiationSpec:
    """Spec for a negotiation scenario."""

    description: str
    environment_description: str
    initial_state_description: str
    hidden_preferences: dict[str, Any]  # priorities, reservation, aspiration, batna
    max_rounds: int
    success_criteria: list[str]
    failure_modes: list[str]
    actions: list[SimulationActionSpecModel]
    max_steps: int = 0  # auto-derived from max_rounds * 2 if not set

    def __post_init__(self) -> None:
        if self.max_steps <= 0:
            self.max_steps = max(self.max_rounds * 2, 4)
